lbrynet/p2p/HTTPBlobDownloader.py

```diff
@@ -121,6 +122,12 @@ def _download_and_get_retry_delay(self):
 
     @defer.inlineCallbacks
     def _download_blob(self, blob):
+        """ Given a `BlobFile` object, attempt to download the given
+        blob from any of the given mirrors. up to self.max_failures times.
+
+        :param BlobFile blob: The BlobFile object which we want to download
+        :return:
+        """
```

Comment:
Instead of adding the comment, can you add type annotations to the method itself.

Reply:
I can add both, Sphynx docstrings are much easier to read when docstrings are rendered with reStructuredText, which is why I kept it despite its redundancy.

Reply:
I don't think we're using sphynx doc generator. I imagine that if we do use a doc generator we'd want one that can understand Python 3 style type annotations though.

Reply:
Sphynx does understand Python 3 type annotations. The above isn't supposed to be a good example, just something I wrote for myself and figured wouldn't do any harm if included. 

That being said, Sphynx makes the parameters very easy to read when being rendered.
For something like the following block:
    ```python
    def get_blob(self, blob_hash, length=None):
        """Return a blob identified by blob_hash, which may be a new blob or a
        blob that is already on the hard disk

        :param str blob_hash: String representing the desired blob's hash
        :param length: Length of the blob to download. If None is specified, then
          the blob is assumed to already be saved in the `self.blobs` dict.
        :raises TypeError: If the blob isn't saved in `self.blobs` and `length` is
          passed in as a non-`int` type (`None` by default)
        """
    ```
The docstring would render as the following:
![image](https://user-images.githubusercontent.com/16809287/48681334-698ff100-eb67-11e8-916e-32f3bc9d5287.png)

Although the typing of parameters in between `:param type parameter_name: Description` can be hard to read if you're not rendering. For that reason, most people prefer the following:
    ```python
    def get_blob(self, blob_hash, length=None):
        """Return a blob identified by blob_hash, which may be a new blob or a
        blob that is already on the hard disk

        :param blob_hash: String representing the desired blob's hash
        :type blob_hash: str
        :param length: Length of the blob to download. If None is specified, then
          the blob is assumed to already be saved in the `self.blobs` dict.
        :type length: int or None
        :raises TypeError: If the blob isn't saved in `self.blobs` and `length` is
          passed in as a non-`int` type (`None` by default)
        """
    ```
The rendering would still be the same:
![image](https://user-images.githubusercontent.com/16809287/48681502-62b5ae00-eb68-11e8-8cb5-20d2d4fb9d23.png)

Reply:
We are using [mypy](http://mypy-lang.org/) on [torba](https://github.com/lbryio/torba) and plan to use it in lbry repo very soon as well. This means using Python 3 style type annotations. If we use Python 3 style type annotations and also add sphynx doc strings with the same information it leads to huge technical debt.

Reply:
That's not correct, `mypy` statically analyzes the entire code base and enforces that the types are what you say they are. It's not just "typehinting". We use it on torba already which is the wallet code for lbrynet (i gave links to mypy and torba in previous comment).

Reply:
We are starting to use it in lbrynet, but it's going to take some time to get all of the code updated before we can start enforcing it, this is a good example of a file that uses lots of type annotations: https://github.com/lbryio/lbry/blob/master/lbrynet/extras/wallet/transaction.py

Reply:
I'll definitely use it, but it doesn't change the fact that a lot of this code is already complicated, and rewriting something that wasn't written well is much harder logistically, than to simply add a docstring and fix it when you have the time to. Do you know if there's a UML diagram for the flow of the project? If not, there should be some made going forward, it would probably allow us to spot redundancies in design before anyone has to implement it.

Reply:
Better to just fix it right away as you're learning it.

Haven't had time for UML diagrams, by the time a UML diagram is created the code will probably change and the UML diagram will be out of date.

Less code = quicker to understand = quicker to change = quicker to innovate and move forward.

The more docs, UML diagrams and other stuff you add the more technical debt you have and the more bogged down you get. Now making a simple code change requires also updating docs and updating UML diagram, etc. Or more likely, people will only make the code change and now the docs and UML diagram are out of date.

It's much more efficient to write clear and tight code that's easy to understand and easy to modify.

Another really good way to document an API is to write readable unit tests, if they clearly show inputs/outputs it does two things: 1) provide working examples and 2) is "living" documentation because if the code changes the tests have to change, otherwise tests fail. This is *much* better than static written docs which get stale when code changes.

Reply:
If you're moving forward all the time and not looking ahead, you lose sight of the big picture. Is it tedious to draw out a diagram of what the system is? Maybe, but at least when you have one done, you realize "oh well we can get rid of Y and Z because they do the same thing as X, just slightly differently". 

It's good practice to move fast and break stuff, but if you're not thinking about what you're doing then you're just running around aimlessly. While it's great if you can reduce everything down into a purely functional format, that's not always going to be the case, and not everyone will follow that. Hence why you should take out a minute to document what your function is doing so you can think "well huh, this is overly complicated, I can make it simpler"

Reply:
We do brain storming on a white board, this includes sketching stuff roughly resembling UML diagram but much less formal. We did this for the new wallet architecture. But the final design changed enough that the original drawn diagrams don't really reflect reality.

I'm a visual person so totally support quick sketches and diagrams. I just don't want to waste time on formal UML diagrams with expectations that they will be kept updated, etc.

Reply:
I completely understand that. Unfortunately I've never even seen the original so I proposed the UML thing as it would help anyone who doesn't have the privilege of seeing that in person.

Reply:
If we do anything like that again I'll be sure to post a photo into slack.